### PR TITLE
RUMM-3320 [V1] Fix APM local spans not correlating with RUM views

### PR DESCRIPTION
**Pull request by plousada**
### What and why?

APM <-> RUM correlation was not working for local traces due to mismatch in span tag format between the SDK and APM Backend.

fix: https://github.com/DataDog/dd-sdk-ios/issues/1292

### How?

This PR removes the direct use of `RUMContextAttributes` in both Log and Span encoding. 
Instead, a mapping layer is added that transforms keys into the required format for each product. 

On the Example app, changes are made to allow enabling a simple debugging option. It's now possible to single a single log, a single span, and a single traced request on a particular RUM session.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests
- [ ] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: Tests/DatadogTests/Datadog/LoggerTests.swift [hunk 1]
```diff
@@ -562,7 +562,7 @@ class LoggerTests: XCTestCase {
 
         logMatchers.forEach {
             $0.assertValue(
-                forKeyPath: RUMContextAttributes.IDs.sessionID,
+                forKeyPath: "application_id",
```
Comment (plousada) [inline Tests/DatadogTests/Datadog/LoggerTests.swift:565]: This is currently enforcing the schema used by logs backend. 
I'm wondering if it would be preferable to use the output of the mapping function `mapInternalAttributeKey` here instead of the actual expected key and leave that enforce to E2E tests.

File: Tests/DatadogTests/Datadog/WebView/WKUserContentController+DatadogTests.swift [hunk 2]
```diff
@@ -180,7 +180,7 @@ class WKUserContentController_DatadogTests: XCTestCase {
         )
 
         let messageHandler = try XCTUnwrap(controller.messageHandlers.first?.handler) as? DatadogMessageHandler
-        let webLogMessage = MockScriptMessage(body: #"{"eventType":"log","event":{"date":1635932927012,"error":{"origin":"console"},"message":"console error: error","session_id":"0110cab4-7471-480e-aa4e-7ce039ced355","status":"error","view":{"referrer":"","url":"https://datadoghq.dev/browser-sdk-test-playground"}},"tags":["browser_sdk_version:3.6.13"]}"#)
+        let webLogMessage = MockScriptMessage(body: #"{"eventType":"log","event":{"date":1635932927012,"error":{"origin":"console"},"message":"console error: error","session.id":"0110cab4-7471-480e-aa4e-7ce039ced355","status":"error","view":{"referrer":"","url":"https://datadoghq.dev/browser-sdk-test-playground"}},"tags":["browser_sdk_version:3.6.13"]}"#)
```
Comment (fuzzybinary) [inline Tests/DatadogTests/Datadog/WebView/WKUserContentController+DatadogTests.swift:183]: Was this mock message made to be accurate to what the Browser SDK sends us? If so we probably shouldn't change it and verify we still send the proper attributes.

File: Sources/Datadog/Tracing/TracingV2Configuration.swift [hunk 3]
```diff
@@ -72,7 +72,7 @@ internal struct TracingMessageReceiver: FeatureMessageReceiver {
     ///
     /// - Parameter context: The updated core context.
     private func update(context: DatadogContext) -> Bool {
-        if let attributes: [String: String] = context.featuresAttributes["rum"]?.ids {
+        if let attributes: [String: String?] = context.featuresAttributes["rum"]?.ids {
```
Comment (fuzzybinary) [inline Sources/Datadog/Tracing/TracingV2Configuration.swift:75]: I'm curious why this change was needed?

**Review by fuzzybinary — APPROVED** [review]

File: Tests/DatadogTests/Datadog/WebView/WKUserContentController+DatadogTests.swift (hunk 2 above)
Comment (plousada) [reply Tests/DatadogTests/Datadog/WebView/WKUserContentController+DatadogTests.swift:183]: You're right, my initial approach doesn't address logs coming from Browser SDK correctly.
Thanks for raising it.

File: Sources/Datadog/RUM/Integrations/RUMContextAttributes.swift [hunk 4]
```diff
@@ -10,11 +10,11 @@ import Foundation
 internal enum RUMContextAttributes {
     internal enum IDs {
         /// The ID of RUM application (`String`).
-        internal static let applicationID = "application_id"
+        internal static let applicationID = "application.id"
 
         /// The ID of current RUM session (standard UUID `String`, lowercased).
         /// In case the session is rejected (not sampled), RUM context is set to empty (`[:]`) in core.
-        internal static let sessionID = "session_id"
+        internal static let sessionID = "session.id"
```
Comment (ncreated) [inline Sources/Datadog/RUM/Integrations/RUMContextAttributes.swift:17]: blocker 🚫/ This change break Session Replay, which expects `"application_id"` and `"session_id"`. With being unable to recognise these new values SR thinks there is no RUM, so it doesn't record anything.

1. We need to update these values [here](https://github.com/DataDog/dd-sdk-ios/blob/10135195869d7f785f7a63201e4ecd84795b0490/DatadogSessionReplay/Sources/DatadogCoreIntegration/RUMContextReceiver.swift#L14-L16).
2. Also please update these values in comments of RUM dependency definition within SR module [here](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-3320/GH-1292-fix-apm-with-rum/DatadogSessionReplay/Sources/DatadogCoreIntegration/RUMDependency.swift) - this definition is a central place where SR defines everything it expects or provides to other modules in SDK context.

I bump the need for 2 things:
- Integration Tests for SR: it is now possible on `feature/v2` - I created [REPLAY-1863](https://datadoghq.atlassian.net/browse/REPLAY-1863)
- Integration Unit Tests: to validate feature's contract in unit tests - @maxep do we have JIRA for this?

File: Sources/Datadog/Logging/LoggingV2Configuration.swift [hunk 5]
```diff
@@ -27,6 +27,17 @@ internal func createLoggingConfiguration(
...
+internal func mapInternalAttributeKey(_ originalAttribute: String) -> String {
+    switch originalAttribute {
+    case "application.id":
+        return "application_id"
+    case "session.id":
+        return "session_id"
+    default:
+        return originalAttribute
+    }
+}
```
Comment (ncreated) [inline Sources/Datadog/Logging/LoggingV2Configuration.swift:39]: suggestion/ Could we declare constants for `"application.id"` and `"session.id"` as part of `RUMDependency` namespace defined in `Logging` module? Similar to how we define [`RUMDependency`](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-3320/GH-1292-fix-apm-with-rum/DatadogSessionReplay/Sources/DatadogCoreIntegration/RUMDependency.swift) in Session Replay - single place with listing all attributes that features exchange between each other.

cc @maxep

File: Sources/Datadog/Tracing/TracingV2Configuration.swift [hunk 6]
```diff
@@ -18,6 +18,21 @@ internal func createTracingConfiguration(intake: URL) -> DatadogFeatureConfigura
...
+    case "session.id":
+        return "_dd.session.id"
+    case "view.id":
+        return "_dd.view.id"
+    case "user_action.id":
+        return "_dd.action.id"
+    default:
+        return originalTag
+    }
+}
```
Comment (ncreated) [inline Sources/Datadog/Tracing/TracingV2Configuration.swift:34]: suggestion/ Similar ask to the one in Logging:
> suggestion/ Could we declare constants for `"application.id"` and `"session.id"` as part of `RUMDependency` namespace defined in `Tracing` module? Similar to how we define [`RUMDependency`](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-3320/GH-1292-fix-apm-with-rum/DatadogSessionReplay/Sources/DatadogCoreIntegration/RUMDependency.swift) in Session Replay - single place with listing all attributes that features exchange between each other.

cc @maxep

**Review by ncreated — CHANGES_REQUESTED** [review]

File: Sources/Datadog/Logging/LoggingV2Configuration.swift (hunk 5 above)
Comment (ncreated) [reply Sources/Datadog/Logging/LoggingV2Configuration.swift:39]: Because we work with V1 code (no modules yet), it would have to be `enum LogsRUMDependency` (or similar) to scope it properly.

File: Sources/Datadog/RUM/Integrations/RUMContextAttributes.swift (hunk 4 above)
Comment (maxep) [reply Sources/Datadog/RUM/Integrations/RUMContextAttributes.swift:17]: > Integration Unit Tests: to validate feature's contract in unit tests - @maxep do we have JIRA for this?

File: Sources/Datadog/Logging/LoggingV2Configuration.swift (hunk 5 above)
Comment (maxep) [reply Sources/Datadog/Logging/LoggingV2Configuration.swift:39]: Having constants makes sense, should be defined alongside the message receiver IMO.

File: Sources/Datadog/Tracing/TracingV2Configuration.swift (hunk 6 above)
Comment (ganeshnj) [reply Sources/Datadog/Tracing/TracingV2Configuration.swift:34]: If the constants are shared eg "application_id", declaring them each module (Tracing, Session Replay etc) seems error prone. IMO, it should go atleast in Internal module which is shared by all other modules.

Not related to this, but having an API package for each feature makes these things easier to maintain, where there could be `RUMApi` module with plain structs, APIs and constants (ideal world) which can be taken as dependency by others without all the baggage of RUM.

File: Sources/Datadog/Logging/LoggingV2Configuration.swift (hunk 5 above)
Comment (ncreated) [reply Sources/Datadog/Logging/LoggingV2Configuration.swift:39]: Message receiver can be a good place to define "received" attributes. Where to define "provided" attributes?

Looking at mentioned [RUMDependency](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-3320/GH-1292-fix-apm-with-rum/DatadogSessionReplay/Sources/DatadogCoreIntegration/RUMDependency.swift) in SR there are two parts:
```swift
internal enum RUMDependency {
   // MARK: Contract from RUM to SR:
   static let rumBaggageKey = "rum"
   static let serverTimeOffsetKey = "server_time_offset"
   static let ids = "ids"

   // MARK: Contract from SR to RUM (mirror of `SessionReplayDependency` in RUM):
   static let srBaggageKey = "session-replay"
   static let hasReplay = "has_replay"
}
```
and I believe it is good and clear enough - single place describing the whole contract.

If we want to separate it, it's OK for me as long as it is clear and consistent across all modules.

Comment (maxep) [reply Sources/Datadog/Logging/LoggingV2Configuration.swift:39]: > and I believe it is good and clear enough - single place describing the whole contract.

Yep, totally. Just to add my 2 cents: I don't think we should enforce any convention across modules on that. Some Features will use/share a lot of context values, some won't. We can have some flexibility on how Features are implemented internally IMHO

**Review by ncreated — APPROVED** [review]
@maxep || @ganeshnj I pushed this topic forward and now it is ready for review. 
* Only fixing Session Replay compared to previous work (the change here has low impact as we don't provide SR in V1 except private beta branch, but I did it anyway). 
* Like I mentioned on Slack, this is only minimal change to make it work. 
* We're fixing V1 here - V2 fix is in #1375 
* This PR is targeting `release/1.22.0` branch - I will do release after merge.

CI is green, Bitrise checks are false:

Screenshot 2023-07-18 at 14 11 59

File: Datadog/Example/Debugging/DebugRUMSessionViewController.swift [hunk 7]
```diff
@@ -39,6 +39,12 @@ private class DebugRUMSessionViewModel: ObservableObject {
     @Published var errorMessage: String = ""
     @Published var resourceKey: String = ""
 
+    @Published var logMessage: String = ""
+    @Published var spanOperationName: String = ""
+    @Published var instrumentedRequestURL: String = "https://api.shopist.io/checkout.json"
```
Comment (ganeshnj) [inline Datadog/Example/Debugging/DebugRUMSessionViewController.swift:44]: currently it returns 
```json
{
    "status": 404,
    "error": "Not Found"
}
```

expected ?

**Review by ganeshnj — APPROVED** [review]

**Review by maxep — APPROVED** [review]
Looks great! I left a question.

File: Sources/Datadog/Tracing/Span/SpanEventEncoder.swift [hunk 8]
```diff
@@ -243,8 +243,10 @@ internal struct SpanEventEncoder {
         }
 
         try span.tags.forEach {
-            let metaKey = "meta.\($0.key)"
-            try container.encode($0.value, forKey: DynamicCodingKey(metaKey))
+            if $0.value != "null" {
```
[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Tracing/Span/SpanEventEncoder.swift]: No, if context includes `nil` value (e.g. `["user_action.id": nil]`) that was being encoded as `"null"` here. But it is a good call-out. I fixed it now in a better way, by doing this after context retrieval in `TracingV2Configuration.swift`:
```swift
let attributes = attributes.compactMapValues { $0 }
```
(it's analogous to the fix in V2)

File: Datadog/Example/Debugging/DebugRUMSessionViewController.swift [hunk 9]
```diff
@@ -74,6 +80,10 @@ private class DebugRUMSessionViewModel: ObservableObject {
         )
 
         Global.rum.addUserAction(type: .custom, name: actionName)
+
+        Thread.sleep(forTimeInterval: 0.05)
+        sendSpan()
+
         self.actionName = ""
```
[the remark this answers is not in the document]
Comment (ncreated) [reply Datadog/Example/Debugging/DebugRUMSessionViewController.swift]: ✅ Done

File: Datadog/Example/Debugging/DebugRUMSessionViewController.swift (hunk 7 above)
Comment (ncreated) [reply Datadog/Example/Debugging/DebugRUMSessionViewController.swift:44]: It is `POST` endpoint (`GET` always returns `404`), but getting some % of errors on it is expected as this is part of demo environment 👍.
